### PR TITLE
Remove quadratic loops when exiting

### LINKED ISSUE #12712: Mainnet block construction latency increase with full exists
Network: Mainnet
Commit: 7aa043892bf5983ab41c5182fa4346846deb32da
Setup: Mock block construction in every slot. When there are voluntary exits to pack, the block construction latency increases 300ms due to inefficiency in `InitiateValidatorExit` and having to loop through every validator

image

### PR DESCRIPTION
**Pull request by potuz**
fixes #12712

### CODE REVIEW

File: beacon-chain/core/validators/validator.go [hunk 1]
```diff
@@ -13,11 +13,33 @@ import (
...
+				maxExitEpoch = e
+				churn = 1
+			} else if e == maxExitEpoch {
+				churn++
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		logrus.WithError(err).Error("this can't happen")
```
Comment (terencechain) [inline beacon-chain/core/validators/validator.go:38]: Can we remove `this can't happen` or use something else? The description is not very helpful

Comment (potuz) [reply beacon-chain/core/validators/validator.go:38]: this literally can't happen, it's a quirk of the function signature for `ReadFromEveryValidator` that requires an error to be returned, but this function does never return an error

Comment (potuz) [reply beacon-chain/core/validators/validator.go:38]: I could change this to an `assert(err == nil)` or a panic for all I know.

Comment (terencechain) [reply beacon-chain/core/validators/validator.go:38]: I personally think either option is better than a `this can't happen` log

Comment (terencechain) [reply beacon-chain/core/validators/validator.go:38]: or silence the error with a comment that says it can't happen
`_ = err`

Comment (potuz) [reply beacon-chain/core/validators/validator.go:38]: I truly hate linters :) will use this last one

File: beacon-chain/core/epoch/epoch_processing.go [hunk 2]
```diff
@@ -110,7 +110,10 @@ func ProcessRegistryUpdates(ctx context.Context, state state.BeaconState) (state
 		isActive := helpers.IsActiveValidator(validator, currentEpoch)
 		belowEjectionBalance := validator.EffectiveBalance <= ejectionBal
 		if isActive && belowEjectionBalance {
-			state, err = validators.InitiateValidatorExit(ctx, state, primitives.ValidatorIndex(idx))
+			// Here is fine to do a quadratic loop since this should
+			// barely happen
+			maxExitEpoch, churn := validators.ValidatorsMaxExitEpochsAndChurn(state)
```
Comment (terencechain) [inline beacon-chain/core/epoch/epoch_processing.go:115]: Would it be an issue if there's a mass slashing?

Comment (potuz) [reply beacon-chain/core/epoch/epoch_processing.go:115]: why?

File: proto/prysm/v1alpha1/attestation/aggregation/attestations/maxcover_test.go [hunk 3]
```diff
@@ -433,3 +434,15 @@ func TestAggregateAttestations_aggregateAttestations(t *testing.T) {
 		})
 	}
 }
+
+func TestFilterContained(t *testing.T) {
```
Comment (rkapka) [inline proto/prysm/v1alpha1/attestation/aggregation/attestations/maxcover_test.go:438]: This shouldn't be here

File: beacon-chain/core/validators/validator.go [hunk 4]
```diff
@@ -13,11 +13,34 @@ import (
...
 	"github.com/prysmaticlabs/prysm/v4/time/slots"
 )
 
+// ValidatorAlreadyExitedErr is an error raised when trying to process an exit of
+// an already exited validator
+var ValidatorAlreadyExitedErr = errors.New("validator already exited")
+
+// ValidatorsMaxExitEpochsAndChurn returns the maximum non-FAR_FUTURE_EPOCH exit
+// epoch and the number of them
+func ValidatorsMaxExitEpochsAndChurn(s state.BeaconState) (maxExitEpoch primitives.Epoch, churn uint64) {
```
Comment (rkapka) [inline beacon-chain/core/validators/validator.go:26]:
```suggestion
func ValidatorsMaxExitEpochAndChurn(s state.BeaconState) (maxExitEpoch primitives.Epoch, churn uint64) {
```

File: beacon-chain/core/blocks/exit.go [hunk 5]
```diff
@@ -61,8 +63,15 @@ func ProcessVoluntaryExits(
...
-		if err != nil {
+		beaconState, exitEpoch, err = v.InitiateValidatorExit(ctx, beaconState, exit.Exit.ValidatorIndex, maxExitEpoch, churn)
+		if err == nil {
+			if exitEpoch > maxExitEpoch {
+				maxExitEpoch = exitEpoch
+				churn = 1
+			} else if exitEpoch == maxExitEpoch {
+				churn++
+			}
+		} else if err != v.ValidatorAlreadyExitedErr {
```
Comment (rkapka) [inline beacon-chain/core/blocks/exit.go:74]: I am not sure that I like silently doing nothing when a validator already exited. We could at least log something.

File: beacon-chain/core/blocks/exit.go [hunk 6]
```diff
@@ -61,8 +63,15 @@ func ProcessVoluntaryExits(
...
 			return nil, errors.Wrapf(err, "could not verify exit %d", idx)
 		}
-		beaconState, err = v.InitiateValidatorExit(ctx, beaconState, exit.Exit.ValidatorIndex)
-		if err != nil {
+		beaconState, exitEpoch, err = v.InitiateValidatorExit(ctx, beaconState, exit.Exit.ValidatorIndex, maxExitEpoch, churn)
+		if err == nil {
+			if exitEpoch > maxExitEpoch {
+				maxExitEpoch = exitEpoch
+				churn = 1
+			} else if exitEpoch == maxExitEpoch {
```
Comment (rkapka) [inline beacon-chain/core/blocks/exit.go:71]: can't we simplify this to `else {`? I don't think `exitEpoch < maxExitEpoch` is ever possible

File: beacon-chain/core/validators/validator.go [hunk 7]
```diff
@@ -13,11 +13,34 @@ import (
...
+	farFutureEpoch := params.BeaconConfig().FarFutureEpoch
+	err := s.ReadFromEveryValidator(func(idx int, val state.ReadOnlyValidator) error {
+		e := val.ExitEpoch()
+		if e != farFutureEpoch {
+			if e > maxExitEpoch {
+				maxExitEpoch = e
+				churn = 1
+			} else if e == maxExitEpoch {
```
Comment (rkapka) [inline beacon-chain/core/validators/validator.go:34]: can we simply this to `else {` without any checks?

File: beacon-chain/core/blocks/exit.go (hunk 5 above)
Comment (potuz) [reply beacon-chain/core/blocks/exit.go:74]: This is the current behavior, there's many reasons to process a voluntary exit of an already exited validator, for example if the validator is slashed after exited, we would call this function anyway.

File: beacon-chain/core/blocks/exit.go (hunk 6 above)
Comment (potuz) [reply beacon-chain/core/blocks/exit.go:71]: I agree, but I wanted to be safe of our future self, it's a pretty cheap check though and it's also descriptive.

File: beacon-chain/core/validators/validator.go (hunk 7 above)
Comment (potuz) [reply beacon-chain/core/validators/validator.go:34]: This one can't because it may happen that `e < maxExitEpoch` here.

**Review by rkapka — DISMISSED** [review]

**Review by terencechain — DISMISSED** [review]

File: beacon-chain/core/validators/validator.go [hunk 8]
```diff
@@ -42,73 +65,43 @@ import (
...
-			exitQueueChurn, mErr = mathutil.Add64(exitQueueChurn, 1)
-			if mErr != nil {
-				return mErr
-			}
-		}
-		return nil
-	})
-	if err != nil {
-		return nil, err
+		return s, validator.ExitEpoch, ValidatorAlreadyExitedErr
```
Comment (nisdas) [inline beacon-chain/core/validators/validator.go:79]: This seems wrong, why are we returning an error here ?

File: beacon-chain/core/blocks/exit.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/core/blocks/exit.go:74]: can we use the `errors.Is` pattern here to check for particular error ? Same for the other cases

**Review by nisdas — APPROVED** [review]

**Review by rauljordan — APPROVED** [review]

**Review by nisdas — APPROVED** [review]

**Pull request merged**